File: riscv32_pipeline/tools/rv32im_iss_coremark.c
#include <ctype.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MEM_SIZE (1024u * 1024u)
#define UART_TX  0xFFFFF080u
#define UART_ST  0xFFFFF084u
#define TIMER    0xFFFFF090u

static uint8_t mem[MEM_SIZE];
static uint32_t xreg[32];
static uint32_t pc;
static uint32_t timer_counter;
static char recent[16];
static char finish_recent[40];

static uint32_t u32(uint64_t v) { return (uint32_t)v; }
static int32_t s32(uint32_t v) { return (int32_t)v; }
static int32_t sext(uint32_t v, int bits) {
    uint32_t m = 1u << (bits - 1);
    return (int32_t)((v ^ m) - m);
}

static uint8_t rb(uint32_t addr) {
    if (addr == UART_ST) return 0;
    if (addr == TIMER) return (uint8_t)timer_counter;
    if (addr >= MEM_SIZE) return 0;
    return mem[addr];
}

static uint16_t rh(uint32_t addr) {
    return (uint16_t)(rb(addr) | ((uint16_t)rb(addr + 1) << 8));
}

static uint32_t rw(uint32_t addr) {
    if (addr == UART_ST) return 0;
    if (addr == TIMER) return timer_counter;
    if (addr + 3 >= MEM_SIZE) return 0;
    return (uint32_t)mem[addr] |
           ((uint32_t)mem[addr + 1] << 8) |
           ((uint32_t)mem[addr + 2] << 16) |
           ((uint32_t)mem[addr + 3] << 24);
}

static void push_uart(uint8_t ch) {
    putchar(ch);
    fflush(stdout);
    memmove(recent, recent + 1, sizeof(recent) - 2);
    recent[sizeof(recent) - 2] = (char)ch;
    recent[sizeof(recent) - 1] = 0;
    memmove(finish_recent, finish_recent + 1, sizeof(finish_recent) - 2);
    finish_recent[sizeof(finish_recent) - 2] = (char)ch;
    finish_recent[sizeof(finish_recent) - 1] = 0;
}

static void wb(uint32_t addr, uint32_t val) {
    if (addr == UART_TX) {
        push_uart((uint8_t)val);
        return;
    }
    if (addr < MEM_SIZE) mem[addr] = (uint8_t)val;
}

static void wh(uint32_t addr, uint32_t val) {
    wb(addr, val);
    wb(addr + 1, val >> 8);
}

static void ww(uint32_t addr, uint32_t val) {
    if (addr == UART_TX) {
        push_uart((uint8_t)val);
        return;
    }
    if (addr + 3 >= MEM_SIZE) return;
    mem[addr] = (uint8_t)val;
    mem[addr + 1] = (uint8_t)(val >> 8);
    mem[addr + 2] = (uint8_t)(val >> 16);
    mem[addr + 3] = (uint8_t)(val >> 24);
}

static uint32_t divs(uint32_t a, uint32_t b) {
    int32_t sa = s32(a), sb = s32(b);
    if (sb == 0) return 0xFFFFFFFFu;
    if (sa == (int32_t)0x80000000u && sb == -1) return 0x80000000u;
    return (uint32_t)(sa / sb);
}

static uint32_t rems(uint32_t a, uint32_t b) {
    int32_t sa = s32(a), sb = s32(b);
    if (sb == 0) return a;
    if (sa == (int32_t)0x80000000u && sb == -1) return 0;
    return (uint32_t)(sa % sb);
}

static void load_mem(const char *path) {
    FILE *f = fopen(path, "rb");
    if (!f) {
        fprintf(stderr, "cannot open %s\n", path);
        exit(2);
    }
    char line[256];
    uint32_t addr = 0;
    while (fgets(line, sizeof(line), f)) {
        char *p = line;
        while (isspace((unsigned char)*p)) p++;
        if (*p == 0 || *p == '/' || *p == '#') continue;
        if (*p == '@') {
            addr = (uint32_t)strtoul(p + 1, NULL, 16) * 4u;
            continue;
        }
        uint32_t word = (uint32_t)strtoul(p, NULL, 16);
        if (addr + 3 < MEM_SIZE) {
            mem[addr] = (uint8_t)word;
            mem[addr + 1] = (uint8_t)(word >> 8);
            mem[addr + 2] = (uint8_t)(word >> 16);
            mem[addr + 3] = (uint8_t)(word >> 24);
        }
        addr += 4;
    }
    fclose(f);
}

static int step(void) {
    uint32_t insn = rw(pc);
    uint32_t pc0 = pc;
    pc = pc + 4;
    uint32_t op = insn & 0x7f;
    uint32_t rd = (insn >> 7) & 31;
    uint32_t f3 = (insn >> 12) & 7;
    uint32_t rs1 = (insn >> 15) & 31;
    uint32_t rs2 = (insn >> 20) & 31;
    uint32_t f7 = (insn >> 25) & 0x7f;
    uint32_t imm, addr, sh;

    switch (op) {
    case 0x37:
        xreg[rd] = insn & 0xFFFFF000u;
        break;
    case 0x17:
        xreg[rd] = pc0 + (insn & 0xFFFFF000u);
        break;
    case 0x6f:
        imm = ((insn >> 31) << 20) | (((insn >> 12) & 0xff) << 12) |
              (((insn >> 20) & 1) << 11) | (((insn >> 21) & 0x3ff) << 1);
        xreg[rd] = pc;
        pc = pc0 + sext(imm, 21);
        break;
    case 0x67:
        imm = (uint32_t)sext(insn >> 20, 12);
        addr = (xreg[rs1] + imm) & ~1u;
        xreg[rd] = pc;
        pc = addr;
        break;
    case 0x63: {
        imm = ((insn >> 31) << 12) | (((insn >> 7) & 1) << 11) |
              (((insn >> 25) & 0x3f) << 5) | (((insn >> 8) & 0xf) << 1);
        int take = 0;
        if (f3 == 0) take = xreg[rs1] == xreg[rs2];
        else if (f3 == 1) take = xreg[rs1] != xreg[rs2];
        else if (f3 == 4) take = s32(xreg[rs1]) < s32(xreg[rs2]);
        else if (f3 == 5) take = s32(xreg[rs1]) >= s32(xreg[rs2]);
        else if (f3 == 6) take = xreg[rs1] < xreg[rs2];
        else if (f3 == 7) take = xreg[rs1] >= xreg[rs2];
        if (take) pc = pc0 + sext(imm, 13);
        break;
    }
    case 0x03:
        addr = xreg[rs1] + (uint32_t)sext(insn >> 20, 12);
        if (f3 == 0) xreg[rd] = (uint32_t)sext(rb(addr), 8);
        else if (f3 == 1) xreg[rd] = (uint32_t)sext(rh(addr), 16);
        else if (f3 == 2) xreg[rd] = rw(addr);
        else if (f3 == 4) xreg[rd] = rb(addr);
        else if (f3 == 5) xreg[rd] = rh(addr);
        break;
    case 0x23:
        imm = ((insn >> 25) << 5) | rd;
        addr = xreg[rs1] + (uint32_t)sext(imm, 12);
        if (f3 == 0) wb(addr, xreg[rs2]);
        else if (f3 == 1) wh(addr, xreg[rs2]);
        else if (f3 == 2) ww(addr, xreg[rs2]);
        break;
    case 0x13:
        imm = (uint32_t)sext(insn >> 20, 12);
        sh = (insn >> 20) & 31;
        if (f3 == 0) xreg[rd] = xreg[rs1] + imm;
        else if (f3 == 2) xreg[rd] = s32(xreg[rs1]) < (int32_t)imm;
        else if (f3 == 3) xreg[rd] = xreg[rs1] < imm;
        else if (f3 == 4) xreg[rd] = xreg[rs1] ^ imm;
        else if (f3 == 6) xreg[rd] = xreg[rs1] | imm;
        else if (f3 == 7) xreg[rd] = xreg[rs1] & imm;
        else if (f3 == 1) xreg[rd] = xreg[rs1] << sh;
        else if (f3 == 5) xreg[rd] = (f7 == 0x20) ? (uint32_t)(s32(xreg[rs1]) >> sh) : (xreg[rs1] >> sh);
        break;
    case 0x33:
        if (f7 == 1) {
            uint32_t a = xreg[rs1], b = xreg[rs2];
            int64_t sa = s32(a), sb = s32(b);
            uint64_t ua = a, ub = b;
            if (f3 == 0) xreg[rd] = (uint32_t)(a * b);
            else if (f3 == 1) xreg[rd] = (uint32_t)((sa * sb) >> 32);
            else if (f3 == 2) xreg[rd] = (uint32_t)((sa * (int64_t)ub) >> 32);
            else if (f3 == 3) xreg[rd] = (uint32_t)((ua * ub) >> 32);
            else if (f3 == 4) xreg[rd] = divs(a, b);
            else if (f3 == 5) xreg[rd] = b ? (a / b) : 0xFFFFFFFFu;
            else if (f3 == 6) xreg[rd] = rems(a, b);
            else if (f3 == 7) xreg[rd] = b ? (a % b) : a;
        } else {
            sh = xreg[rs2] & 31;
            if (f3 == 0) xreg[rd] = (f7 == 0x20) ? (xreg[rs1] - xreg[rs2]) : (xreg[rs1] + xreg[rs2]);
            else if (f3 == 1) xreg[rd] = xreg[rs1] << sh;
            else if (f3 == 2) xreg[rd] = s32(xreg[rs1]) < s32(xreg[rs2]);
            else if (f3 == 3) xreg[rd] = xreg[rs1] < xreg[rs2];
            else if (f3 == 4) xreg[rd] = xreg[rs1] ^ xreg[rs2];
            else if (f3 == 5) xreg[rd] = (f7 == 0x20) ? (uint32_t)(s32(xreg[rs1]) >> sh) : (xreg[rs1] >> sh);
            else if (f3 == 6) xreg[rd] = xreg[rs1] | xreg[rs2];
            else if (f3 == 7) xreg[rd] = xreg[rs1] & xreg[rs2];
        }
        break;
    case 0x0f:
        break;
    case 0x73:
        return 0;
    default:
        fprintf(stderr, "\nunsupported pc=0x%08x insn=0x%08x op=0x%02x\n", pc0, insn, op);
        return 0;
    }
    xreg[0] = 0;
    timer_counter++;
    return 1;
}

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "usage: %s coremark.mem [max_steps]\n", argv[0]);
        return 2;
    }
    uint64_t max_steps = (argc >= 3) ? strtoull(argv[2], NULL, 0) : 2000000000ull;
    load_mem(argv[1]);
    for (uint64_t i = 0; i < max_steps; i++) {
        if (!step()) break;
        if (strstr(finish_recent, "Finished Successfully")) break;
    }
    printf("\n\n[ISS-C] pc=0x%08x timer=%u\n", pc, timer_counter);
    return 0;
}

File: riscv32_pipeline/tools/rv32im_iss_coremark.py
#!/usr/bin/env python3
import argparse
from pathlib import Path


MASK32 = 0xFFFFFFFF


def u32(x):
    return x & MASK32


def s32(x):
    x &= MASK32
    return x - 0x100000000 if x & 0x80000000 else x


def sext(x, bits):
    sign = 1 << (bits - 1)
    return (x & (sign - 1)) - (x & sign)


def load_mem_words(path, mem):
    addr = 0
    for raw in Path(path).read_text(encoding="utf-8", errors="ignore").splitlines():
        line = raw.strip()
        if not line or line.startswith("//"):
            continue
        if line.startswith("@"):
            addr = int(line[1:], 16) * 4
            continue
        word = int(line.split()[0], 16) & MASK32
        if addr + 4 > len(mem):
            mem.extend(b"\x00" * (addr + 4 - len(mem)))
        mem[addr:addr + 4] = word.to_bytes(4, "little")
        addr += 4


class RV32IM:
    def __init__(self, mem_path, cpu_hz=75_000_000, max_steps=100_000_000):
        self.mem = bytearray(1024 * 1024)
        load_mem_words(mem_path, self.mem)
        self.x = [0] * 32
        self.pc = 0
        self.cpu_hz = cpu_hz
        self.timer = 0
        self.max_steps = max_steps
        self.uart = []
        self.running = True

    def rb(self, addr):
        addr &= MASK32
        if addr == 0xFFFFF084:
            return 0
        if addr == 0xFFFFF090:
            return self.timer & 0xFF
        if addr >= len(self.mem):
            return 0
        return self.mem[addr]

    def rh(self, addr):
        return self.rb(addr) | (self.rb(addr + 1) << 8)

    def rw(self, addr):
        addr &= MASK32
        if addr == 0xFFFFF084:
            return 0
        if addr == 0xFFFFF090:
            return self.timer
        if addr + 4 > len(self.mem):
            return 0
        return int.from_bytes(self.mem[addr:addr + 4], "little")

    def wb(self, addr, val):
        addr &= MASK32
        val &= 0xFF
        if addr == 0xFFFFF080:
            self.uart.append(chr(val))
            print(chr(val), end="", flush=True)
            return
        if addr >= len(self.mem):
            return
        self.mem[addr] = val

    def wh(self, addr, val):
        self.wb(addr, val)
        self.wb(addr + 1, val >> 8)

    def ww(self, addr, val):
        addr &= MASK32
        val &= MASK32
        if addr == 0xFFFFF080:
            self.wb(addr, val)
            return
        if addr >= len(self.mem):
            return
        if addr + 4 > len(self.mem):
            self.mem.extend(b"\x00" * (addr + 4 - len(self.mem)))
        self.mem[addr:addr + 4] = val.to_bytes(4, "little")

    def divs(self, a, b):
        a, b = s32(a), s32(b)
        if b == 0:
            return MASK32
        if a == -0x80000000 and b == -1:
            return 0x80000000
        return u32(int(a / b))

    def rems(self, a, b):
        a, b = s32(a), s32(b)
        if b == 0:
            return u32(a)
        if a == -0x80000000 and b == -1:
            return 0
        return u32(a - int(a / b) * b)

    def step(self):
        insn = self.rw(self.pc)
        pc0 = self.pc
        self.pc = u32(self.pc + 4)
        op = insn & 0x7F
        rd = (insn >> 7) & 0x1F
        f3 = (insn >> 12) & 7
        rs1 = (insn >> 15) & 0x1F
        rs2 = (insn >> 20) & 0x1F
        f7 = (insn >> 25) & 0x7F
        x = self.x

        if op == 0x37:
            x[rd] = u32(insn & 0xFFFFF000)
        elif op == 0x17:
            x[rd] = u32(pc0 + (insn & 0xFFFFF000))
        elif op == 0x6F:
            imm = ((insn >> 31) << 20) | (((insn >> 12) & 0xFF) << 12) | (((insn >> 20) & 1) << 11) | (((insn >> 21) & 0x3FF) << 1)
            x[rd] = self.pc
            self.pc = u32(pc0 + sext(imm, 21))
        elif op == 0x67:
            imm = sext(insn >> 20, 12)
            t = u32(x[rs1] + imm) & ~1
            x[rd] = self.pc
            self.pc = t
        elif op == 0x63:
            imm = ((insn >> 31) << 12) | (((insn >> 7) & 1) << 11) | (((insn >> 25) & 0x3F) << 5) | (((insn >> 8) & 0xF) << 1)
            take = (
                (f3 == 0 and x[rs1] == x[rs2]) or
                (f3 == 1 and x[rs1] != x[rs2]) or
                (f3 == 4 and s32(x[rs1]) < s32(x[rs2])) or
                (f3 == 5 and s32(x[rs1]) >= s32(x[rs2])) or
                (f3 == 6 and x[rs1] < x[rs2]) or
                (f3 == 7 and x[rs1] >= x[rs2])
            )
            if take:
                self.pc = u32(pc0 + sext(imm, 13))
        elif op == 0x03:
            addr = u32(x[rs1] + sext(insn >> 20, 12))
            if f3 == 0:
                x[rd] = u32(sext(self.rb(addr), 8))
            elif f3 == 1:
                x[rd] = u32(sext(self.rh(addr), 16))
            elif f3 == 2:
                x[rd] = self.rw(addr)
            elif f3 == 4:
                x[rd] = self.rb(addr)
            elif f3 == 5:
                x[rd] = self.rh(addr)
        elif op == 0x23:
            imm = ((insn >> 25) << 5) | rd
            addr = u32(x[rs1] + sext(imm, 12))
            if f3 == 0:
                self.wb(addr, x[rs2])
            elif f3 == 1:
                self.wh(addr, x[rs2])
            elif f3 == 2:
                self.ww(addr, x[rs2])
        elif op == 0x13:
            imm = sext(insn >> 20, 12)
            sh = (insn >> 20) & 0x1F
            if f3 == 0:
                x[rd] = u32(x[rs1] + imm)
            elif f3 == 2:
                x[rd] = 1 if s32(x[rs1]) < imm else 0
            elif f3 == 3:
                x[rd] = 1 if x[rs1] < u32(imm) else 0
            elif f3 == 4:
                x[rd] = u32(x[rs1] ^ imm)
            elif f3 == 6:
                x[rd] = u32(x[rs1] | imm)
            elif f3 == 7:
                x[rd] = u32(x[rs1] & imm)
            elif f3 == 1:
                x[rd] = u32(x[rs1] << sh)
            elif f3 == 5:
                x[rd] = u32(s32(x[rs1]) >> sh) if f7 == 0x20 else u32(x[rs1] >> sh)
        elif op == 0x33:
            if f7 == 1:
                a, b = x[rs1], x[rs2]
                if f3 == 0:
                    x[rd] = u32(a * b)
                elif f3 == 1:
                    x[rd] = u32((s32(a) * s32(b)) >> 32)
                elif f3 == 2:
                    x[rd] = u32((s32(a) * b) >> 32)
                elif f3 == 3:
                    x[rd] = u32((a * b) >> 32)
                elif f3 == 4:
                    x[rd] = self.divs(a, b)
                elif f3 == 5:
                    x[rd] = MASK32 if b == 0 else u32(a // b)
                elif f3 == 6:
                    x[rd] = self.rems(a, b)
                elif f3 == 7:
                    x[rd] = a if b == 0 else u32(a % b)
            else:
                if f3 == 0:
                    x[rd] = u32(x[rs1] - x[rs2]) if f7 == 0x20 else u32(x[rs1] + x[rs2])
                elif f3 == 1:
                    x[rd] = u32(x[rs1] << (x[rs2] & 31))
                elif f3 == 2:
                    x[rd] = 1 if s32(x[rs1]) < s32(x[rs2]) else 0
                elif f3 == 3:
                    x[rd] = 1 if x[rs1] < x[rs2] else 0
                elif f3 == 4:
                    x[rd] = u32(x[rs1] ^ x[rs2])
                elif f3 == 5:
                    sh = x[rs2] & 31
                    x[rd] = u32(s32(x[rs1]) >> sh) if f7 == 0x20 else u32(x[rs1] >> sh)
                elif f3 == 6:
                    x[rd] = u32(x[rs1] | x[rs2])
                elif f3 == 7:
                    x[rd] = u32(x[rs1] & x[rs2])
        elif op == 0x0F:
            pass
        elif op == 0x73:
            self.running = False
        else:
            raise RuntimeError(f"unsupported opcode pc=0x{pc0:08x} insn=0x{insn:08x}")

        x[0] = 0
        self.timer = u32(self.timer + 1)

    def run(self):
        for step in range(self.max_steps):
            if not self.running:
                break
            self.step()
            if "CoreMark 1.0" in "".join(self.uart):
                break
        return "".join(self.uart)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("mem")
    ap.add_argument("--max-steps", type=int, default=100_000_000)
    args = ap.parse_args()
    sim = RV32IM(args.mem, max_steps=args.max_steps)
    out = sim.run()
    print(f"\n\n[ISS] pc=0x{sim.pc:08x} timer={sim.timer} chars={len(out)}")


if __name__ == "__main__":
    main()
